ide/db/src/org/netbeans/modules/db/explorer/ConnectionList.java

```diff
@@ -273,32 +278,57 @@ public void setPreferredConnection(DatabaseConnection c) {
             }
             ref.clear();
             preferred = new PreferredRef(c);
+            lastPrefName = null;
         }
         dbPreferences.put(PREF_PREFERRED_CONNECTION_NAME, c.getName());
         fireListeners();
     }
     
     public DatabaseConnection getPreferredConnection(boolean selectFirst) {
-        DatabaseConnection p = preferred.get();
-        if (p != null) {
-            return p;
+        String fallback;
+        DatabaseConnection p;
+        synchronized (this) {
+            p = preferred.get();
+            if (p != null) {
+                return p;
+            }
+            fallback = this.lastPrefName;
         }
         String prefName = dbPreferences.get(PREF_PREFERRED_CONNECTION_NAME, null);
         DatabaseConnection[] conns = getConnections();
-        if (prefName == null) {
-            return conns.length == 0 || !selectFirst ? null : conns[0];
+        DatabaseConnection selected = findConnection(prefName, conns);
+        boolean prefChanged = false;
+        if (selected == null) {
+            if (conns.length > 0) {
+                // find the last one, or just anything.
+                selected = findConnection(fallback, conns);
+            }
+            if (selected != null) {
+                prefChanged = !Objects.equals(lastPrefName, selected.getName());
+                lastPrefName = selected.getName();
+            } else {
+                return null;
+            }
         }
+        synchronized (this) {
+            p = preferred.get();
+            if (p == null) {
+                preferred = new PreferredRef(selected);
+                p = selected;
+            }
+        }
+        if (prefChanged) {
+            fireListeners();
+        }
+        return p;
+    }
+    
+    private static DatabaseConnection findConnection(String name, DatabaseConnection[] conns) {
         for (int i = 0; i < conns.length; i++) {
```

Comment:
for-each loop maybe?

Reply:
I don't see the problem, it's a matter of personal preference.

Reply:
I usually use for-loops in these situations ... this part was actually extracted from the older code that is obviously no excuse for not updating the style. I usually incorporate 'style issues' whenever some func/perf ("real") issue is found and a new commit has to be made.